keanu-project/src/test/java/io/improbable/keanu/e2e/regression/RadonHierarchicalRegression.java

```diff
@@ -1,181 +1,213 @@
-        double[] radonForSubModel = Arrays.copyOfRange(allRadon, startIndex, endIndex);
+        double[] floorForSubModel = data.stream().mapToDouble(d -> d.floor).toArray();
+        double[] radonForSubModel = data.stream().mapToDouble(d -> d.log_radon).toArray();
 
         DoubleTensor x = DoubleTensor.create(floorForSubModel, floorForSubModel.length, 1);
         DoubleTensor y = DoubleTensor.create(radonForSubModel, floorForSubModel.length, 1);
 
-        RegressionModel model = RegressionModel.
+        return RegressionModel.
             withTrainingData(x, y).
             withRegularization(RegressionRegularization.RIDGE).
             withPriorOnWeights(muGradient, sigmaGradient).
             withPriorOnIntercept(muIntercept, sigmaIntercept).
             buildWithoutFitting();
+    }
 
-        return model;
+    private NetworkSamples NUTSSample(BayesianNetwork bayesianNetwork) {
+        Vertex muAlpha = bayesianNetwork.getVertexByLabel(new VertexLabel("MuIntercept"));
+        Vertex sigmaAlpha = bayesianNetwork.getVertexByLabel(new VertexLabel("SigmaIntercept"));
+        Vertex muBeta = bayesianNetwork.getVertexByLabel(new VertexLabel("MuGradient"));
+        Vertex sigmaBeta = bayesianNetwork.getVertexByLabel(new VertexLabel("SigmaGradient"));
+
+        // note that way too few samples are taken due to time constraints
+        KeanuProbabilisticModelWithGradient probabilisticModel = new KeanuProbabilisticModelWithGradient(bayesianNetwork);
+        return NUTS.builder()
+            .maxTreeHeight(5)
+            .adaptEnabled(true)
+            .build()
+            .getPosteriorSamples(probabilisticModel, Arrays.asList(muAlpha, muBeta, sigmaAlpha, sigmaBeta), 100)
+            .downSample(2);
     }
 
-    private void optimise(BayesianNetwork bayesianNetwork, List<RegressionModel> models) {
-        bayesianNetwork.probeForNonZeroProbability(100);
+    private void optimise(BayesianNetwork bayesianNetwork) {
         GradientOptimizer optimizer = KeanuOptimizer.Gradient.builderFor(bayesianNetwork)
-            .absoluteThreshold(0.25)
             .maxEvaluations(10000)
             .build();
         optimizer.maxAPosteriori();
-
-        assertValuesAreCorrect(bayesianNetwork, models);
     }
 
-    private void assertValuesAreCorrect(BayesianNetwork bayesianNetwork, List<RegressionModel> models) {
-        DoubleVertex muIntercept = (DoubleVertex) bayesianNetwork.getVertexByLabel(new VertexLabel("MuIntercept"));
-        DoubleVertex muGradient = (DoubleVertex) bayesianNetwork.getVertexByLabel(new VertexLabel("MuGradient"));
-        DoubleVertex sigmaIntercept = (DoubleVertex) bayesianNetwork.getVertexByLabel(new VertexLabel("SigmaIntercept"));
-        DoubleVertex sigmaGradient = (DoubleVertex) bayesianNetwork.getVertexByLabel(new VertexLabel("SigmaGradient"));
+    private void assertSampleAveragesAreCorrect(final GaussianVertex muAlpha,
```

Comment:
Can you run a quick check to make sure that the variance in the samples are correct? Sometimes the sampling algorithms reject every sample if they're in an area of low probability you will see no variance. Would be good to check it is exploring the space somewhat and not staying at its original position!

Reply:
It's good you raised this because it looks like the variance is basically zero.

Obviously the number of samples we are taking is tiny, but if i put it up to 1000 and downsample by 10 the variance is still pretty much 0.